<?xml version="1.0"?>
<launch>
  <arg name="serial_no_d435_right"           default="053122251286"/>  <!-- Right and back are swapped -->
  <arg name="serial_no_d435_left"            default="141222070808"/>
  <arg name="serial_no_d455_front"           default="053122251335"/>
  <arg name="serial_no_d455_back"            default="939622074382"/>
  
  <arg name="initial_reset"                   default="true"/>
  <arg name="sim"                             default="false"/>
  <arg name="width"                           default="424"/>
  <arg name="height"                          default="240"/>
  <arg name="fps"                             default="15"/>

  <arg name="extra_width"                     default="424"/>
  <arg name="extra_height"                    default="240"/>
  <arg name="extra_fps"                       default="15"/>
  <!-- Launch with any combination of front, back, and extra_cams. Front and Back are mapping cams, extra_cams are just vision. -->
  <!-- If only one of the mapping cams is enabled it will use RGB and Depth to map and odom -->
  <arg name="front"                           default="true"/>
  <arg name="back"                            default="true"/>
  <arg name="extra_cams"                      default="true"/>
  <!-- If both it will generate pointclouds on each and combine them and then use that for mapping -->
  <arg name="pointcloud" default="$(eval arg('front') and arg('back'))" />
  
  <arg name="decimation" default="4"/>
  <arg name="voxel_size" default="0.01"/> <!--meters-->
  <arg name="noise_filter_radius" default="0.1"/> <!--meters radius-->
  <arg name="max_depth" default="3"/> <!--meters-->

  <group unless="$(arg sim)">

    
    <!-- Right Camera Temporarily Disabled (since it is the old d455 back, doesn't work)-->
    <!-- <group ns="d435_right/camera" if="$(arg extra_cams)"> -->

      <!-- Launch the camera device nodelet-->
      <!-- <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
        <arg name="external_manager"         value="false"/>
        <arg name="manager"                  value="right_realsense2_camera_manager"/>
        <arg name="tf_prefix"                value="d435_right"/>
        <arg name="serial_no"                value="$(arg serial_no_d435_right)"/>

        <arg name="initial_reset"            value="$(arg initial_reset)"/>

        <arg name="enable_pointcloud"        value="false"/>
        <arg name="enable_sync"              value="false"/>
        <arg name="align_depth"              value="false"/>

        <arg name="enable_depth"             value="false"/>

        <arg name="color_width"              value="$(arg width)"/>
        <arg name="color_height"             value="$(arg height)"/>
        <arg name="enable_color"             value="true"/>

        <arg name="color_fps"                value="$(arg extra_fps)"/>
        <arg name="filters"                  value=""/>

        <arg name="publish_tf"               value="true"/>
        <arg name="tf_publish_rate"          value="0"/>
      </include>
    </group> -->

    <!-- Left Camera -->
    <group ns="d435_left/camera" if="$(arg extra_cams)">
      <!-- Launch the camera device nodelet-->
      <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
        <arg name="external_manager"         value="false"/>
        <arg name="manager"                  value="left_realsense2_camera_manager"/>
        <arg name="tf_prefix"                value="d435_left"/>
        <arg name="serial_no"                value="$(arg serial_no_d435_left)"/>

        <arg name="initial_reset"            value="$(arg initial_reset)"/>


        <arg name="enable_pointcloud"        value="false"/>
        <arg name="enable_sync"              value="false"/>
        <arg name="align_depth"              value="false"/>

        <arg name="enable_depth"             value="false"/>

        <arg name="color_width"              value="$(arg width)"/>
        <arg name="color_height"             value="$(arg height)"/>
        <arg name="enable_color"             value="true"/>

        <arg name="color_fps"                value="$(arg extra_fps)"/>
        <arg name="filters"                  value=""/>

        <arg name="publish_tf"               value="true"/>
        <arg name="tf_publish_rate"          value="0"/>
      </include>
    </group>

  </group>

  <group if="$(arg sim)">
    <group ns="d435_forward/color" if="$(arg front)">
      <node pkg="image_proc" type="image_proc" name="image_proc" />
      <include file="$(find lunabot_perception)/launch/apriltag.launch">
        <arg name="camera_frame" value="d435_forward_link" />
      </include>
    </group>

    <group ns="d435_backward/color" if="$(arg back)">
      <node pkg="image_proc" type="image_proc" name="image_proc" />
      <include file="$(find lunabot_perception)/launch/apriltag.launch">
        <arg name="camera_frame" value="d435_backward_link" />
      </include>
    </group>
  </group>
</launch>
